From f09edb9d7031957d803a00c0bfb4578e9a94be5d Mon Sep 17 00:00:00 2001
From: Philippe Normand <philn@igalia.com>
Date: Fri, 13 May 2011 17:00:10 +0200
Subject: [PATCH 2/2] search: some fixes on top of Seif's patch

---
 js/ui/searchDisplay.js |   28 ++++++++++++++++++++--------
 1 files changed, 20 insertions(+), 8 deletions(-)

diff --git a/js/ui/searchDisplay.js b/js/ui/searchDisplay.js
index 0c4d3bb..6595e19 100644
--- a/js/ui/searchDisplay.js
+++ b/js/ui/searchDisplay.js
@@ -224,9 +224,11 @@ SearchResults.prototype = {
         this._selectedProvider = -1;
         this._providers = this._searchSystem.getProviders();
         this._providerMeta = [];
-        for (let i = 0; i < this._providers.length; i++)
+        this._providerMetaResults = {};
+        for (let i = 0; i < this._providers.length; i++) {
             this.createProviderMeta(this._providers[i]);
-
+            this._providerMetaResults[this.providers[i].title] = [];
+        }
         this._searchProvidersBox = new St.BoxLayout({ style_class: 'search-providers-box' });
         this.actor.add(this._searchProvidersBox);
 
@@ -266,7 +268,6 @@ SearchResults.prototype = {
         let title = new St.Label({ text: provider.name,
                                    style_class: 'dash-search-button-label' });
 
-        button.label_actor = title;
         bin.set_child(title);
         button.set_child(bin);
         provider.actor = button;
@@ -305,6 +306,12 @@ SearchResults.prototype = {
             meta.actor.hide();
         }
     },
+    
+    _clearDisplayForProvider: function(index) {
+        let meta = this._providerMeta[index];
+        meta.resultDisplay.clear();
+        meta.actor.hide();
+    },
 
     reset: function() {
         this._searchSystem.reset();
@@ -329,7 +336,6 @@ SearchResults.prototype = {
     },
 
     _updateResults: function(searchSystem, results) {
-        this._clearDisplay();
         if (results.length == 0) {
             this._statusText.set_text(_("No matching results."));
             this._statusText.show();
@@ -352,10 +358,16 @@ SearchResults.prototype = {
         for (let i = 0; i < results.length; i++) {
             let [provider, providerResults] = results[i];
             if (providerResults.length == 0)
-                  continue;
-            let meta = this._metaForProvider(provider);
-            meta.actor.show();
-            meta.resultDisplay.renderResults(providerResults, terms);
+                this._clearDisplayForProvider(i)
+            else {
+                if (this._providerMetaResults[provider.title] != providerResults) {
+                    this._providerMetaResults[provider.title] = providerResults;
+                    this._clearDisplayForProvider(i);
+                    let meta = this._metaForProvider(provider);
+                    meta.actor.show();
+                    meta.resultDisplay.renderResults(providerResults, terms);
+                }
+            }
         }
 
         if (this._selectedOpenSearchButton == -1)
-- 
1.7.6.rc2

